File: src/lib/wb/cards.ts
import { wbFetch, type WbCtx } from "./client";
import type {
  WbCardUploadItem,
  WbCardListItem,
  WbCardError,
  WbCardErrorListResponse,
} from "./wbtypes";

/**
 * Submit new cards. NOTE: body must be a top-level ARRAY. A 200 with
 * error:false only means "queued" — the nmID is assigned asynchronously.
 */
export async function uploadCards(
  ctx: WbCtx,
  items: WbCardUploadItem[]
): Promise<void> {
  await wbFetch({
    token: ctx.token,
    sandbox: ctx.sandbox,
    method: "POST",
    path: "/content/v2/cards/upload",
    body: items,
    timeoutMs: 40_000,
  });
}

/** Look up a card by the seller's vendorCode (textSearch). */
export async function findCardByVendorCode(
  ctx: WbCtx,
  vendorCode: string
): Promise<WbCardListItem | null> {
  const res = await wbFetch<{ cards: WbCardListItem[] }>({
    token: ctx.token,
    sandbox: ctx.sandbox,
    method: "POST",
    path: "/content/v2/get/cards/list",
    body: {
      settings: {
        sort: { ascending: false },
        filter: { withPhoto: -1, textSearch: vendorCode },
        cursor: { limit: 100 },
      },
    },
  });
  const cards = res.cards ?? [];
  return cards.find((c) => c.vendorCode === vendorCode) ?? null;
}

/**
 * Poll get/cards/list until the card (with nmID) appears, or timeout.
 * Returns the card, or throws with the error-list reason if creation failed.
 */
export async function waitForCard(
  ctx: WbCtx,
  vendorCode: string,
  opts: { timeoutMs?: number; intervalMs?: number; onTick?: (n: number) => void } = {}
): Promise<WbCardListItem> {
  const timeoutMs = opts.timeoutMs ?? 6 * 60_000; // 6 min default
  const intervalMs = opts.intervalMs ?? 12_000;
  const deadline = Date.now() + timeoutMs;
  let tick = 0;

  while (Date.now() < deadline) {
    tick++;
    opts.onTick?.(tick);
    const card = await findCardByVendorCode(ctx, vendorCode);
    if (card && card.nmID) return card;

    // check error list every other tick (fail fast if WB rejected the card)
    // — halves content calls while polling under concurrency
    if (tick % 2 === 0) {
      const err = await findCardError(ctx, vendorCode);
      if (err && err.errors?.length) {
        throw new Error(`WB 拒绝了卡片 ${vendorCode}: ${err.errors.join("; ")}`);
      }
    }
    await sleep(intervalMs);
  }
  // last error check
  const err = await findCardError(ctx, vendorCode);
  if (err && err.errors?.length) {
    throw new Error(`WB 拒绝了卡片 ${vendorCode}: ${err.errors.join("; ")}`);
  }
  throw new Error(
    `等待 nmID 超时（${Math.round(timeoutMs / 60000)} 分钟）。WB 同步可能延迟，可稍后在历史中重试挂图/定价。`
  );
}

export async function listCardErrors(ctx: WbCtx): Promise<WbCardError[]> {
  // Response shape: { data: { items: [{ vendorCodes, errors: {vc: [msg]} }] } }
  const res = await wbFetch<WbCardErrorListResponse>({
    token: ctx.token,
    sandbox: ctx.sandbox,
    method: "POST",
    path: "/content/v2/cards/error/list",
    query: { locale: "ru" },
    body: {},
  });
  const out: WbCardError[] = [];
  for (const item of res.data?.items ?? []) {
    const errs = item.errors ?? {};
    for (const vc of Object.keys(errs)) {
      out.push({ vendorCode: vc, errors: errs[vc] ?? [] });
    }
  }
  return out;
}

async function findCardError(
  ctx: WbCtx,
  vendorCode: string
): Promise<WbCardError | null> {
  try {
    const errors = await listCardErrors(ctx);
    return errors.find((e) => e.vendorCode === vendorCode) ?? null;
  } catch {
    return null;
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

File: src-tauri/src/wb/cards.rs
#![allow(dead_code)]
//! Card create + poll (ported from src/lib/wb/cards.ts).

use crate::state::AppState;
use crate::wb::client::{wb_fetch, WbCtx, WbReq};
use crate::wb::types::WbCardListItem;
use anyhow::{anyhow, Result};
use serde_json::{json, Value};
use std::time::{Duration, Instant};

/// Submit new cards. Body must be a top-level ARRAY. A 200 only means "queued";
/// the nmID is assigned asynchronously.
pub async fn upload_cards(state: &AppState, ctx: &WbCtx, items: Vec<Value>) -> Result<()> {
    wb_fetch(
        state,
        ctx,
        WbReq::post("/content/v2/cards/upload")
            .body(Value::Array(items))
            .timeout(40_000),
    )
    .await?;
    Ok(())
}

/// Look up a card by the seller's vendorCode (textSearch).
pub async fn find_card_by_vendor_code(
    state: &AppState,
    ctx: &WbCtx,
    vendor_code: &str,
) -> Result<Option<WbCardListItem>> {
    let v = wb_fetch(
        state,
        ctx,
        WbReq::post("/content/v2/get/cards/list").body(json!({
            "settings": {
                "sort": { "ascending": false },
                "filter": { "withPhoto": -1, "textSearch": vendor_code },
                "cursor": { "limit": 100 }
            }
        })),
    )
    .await?;
    let cards: Vec<WbCardListItem> = v
        .get("cards")
        .and_then(|c| serde_json::from_value(c.clone()).ok())
        .unwrap_or_default();
    Ok(cards.into_iter().find(|c| c.vendor_code == vendor_code))
}

/// Poll get/cards/list until the card (with nmID) appears, or timeout.
pub async fn wait_for_card<F: Fn(u64)>(
    state: &AppState,
    ctx: &WbCtx,
    vendor_code: &str,
    on_tick: F,
) -> Result<WbCardListItem> {
    let timeout = Duration::from_secs(6 * 60);
    let interval = Duration::from_secs(12);
    let start = Instant::now();
    let mut tick = 0u64;

    while start.elapsed() < timeout {
        tick += 1;
        on_tick(tick);
        if let Some(card) = find_card_by_vendor_code(state, ctx, vendor_code).await? {
            if card.nm_id > 0 {
                return Ok(card);
            }
        }
        // check error list every other tick (fail fast if WB rejected the card)
        if tick % 2 == 0 {
            if let Some(errs) = find_card_error(state, ctx, vendor_code).await {
                if !errs.is_empty() {
                    return Err(anyhow!("WB 拒绝了卡片 {}: {}", vendor_code, errs.join("; ")));
                }
            }
        }
        tokio::time::sleep(interval).await;
    }
    if let Some(errs) = find_card_error(state, ctx, vendor_code).await {
        if !errs.is_empty() {
            return Err(anyhow!("WB 拒绝了卡片 {}: {}", vendor_code, errs.join("; ")));
        }
    }
    Err(anyhow!(
        "等待 nmID 超时（6 分钟）。WB 同步可能延迟，可稍后在历史中重试挂图/定价。"
    ))
}

/// Returns (vendorCode, [error messages]) pairs from cards/error/list.
pub async fn list_card_errors(state: &AppState, ctx: &WbCtx) -> Result<Vec<(String, Vec<String>)>> {
    let v = wb_fetch(
        state,
        ctx,
        WbReq::post("/content/v2/cards/error/list")
            .q("locale", "ru")
            .body(json!({})),
    )
    .await?;
    let mut out = vec![];
    if let Some(items) = v.pointer("/data/items").and_then(|x| x.as_array()) {
        for item in items {
            if let Some(errs) = item.get("errors").and_then(|e| e.as_object()) {
                for (vc, msgs) in errs {
                    let list: Vec<String> = msgs
                        .as_array()
                        .map(|a| {
                            a.iter()
                                .filter_map(|m| m.as_str().map(|s| s.to_string()))
                                .collect()
                        })
                        .unwrap_or_default();
                    out.push((vc.clone(), list));
                }
            }
        }
    }
    Ok(out)
}

async fn find_card_error(state: &AppState, ctx: &WbCtx, vc: &str) -> Option<Vec<String>> {
    match list_card_errors(state, ctx).await {
        Ok(list) => list.into_iter().find(|(v, _)| v == vc).map(|(_, e)| e),
        Err(_) => None,
    }
}
